File: android/app/src/main/java/com/thepiratebrowser/android/PutIoService.java
package com.thepiratebrowser.android;

import org.json.JSONObject;
import org.json.JSONArray;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class PutIoService {
    private static final String API_BASE = "https://api.put.io/v2";

    public DeviceCode requestDeviceCode(String clientId) throws Exception {
        String response = get(
                "https://api.put.io/v2/oauth2/oob/code?app_id="
                        + encode(clientId)
                        + "&client_name="
                        + encode("Pirate Browser Android"),
                null,
                false
        );
        JSONObject json = new JSONObject(response);
        String code = json.optString("code");
        if (code.trim().isEmpty()) {
            throw new IllegalStateException("put.io did not provide a link code.");
        }
        return new DeviceCode(code);
    }

    public String waitForDeviceToken(DeviceCode deviceCode) throws Exception {
        for (int attempt = 0; attempt < 90; attempt++) {
            try {
                String response = get(
                        "https://api.put.io/v2/oauth2/oob/code/"
                                + encode(deviceCode.code),
                        null,
                        true
                );
                String token = new JSONObject(response).optString("oauth_token");
                if (!token.trim().isEmpty()) return token;
            } catch (PendingAuthorization ignored) {
                // The user has not approved the displayed code yet.
            }
            Thread.sleep(2_000);
        }
        throw new IllegalStateException("put.io linking timed out. Try again.");
    }

    public String addTransfer(String oauthToken, String magnet) throws Exception {
        String response = postForm("/transfers/add", oauthToken, "url=" + encode(magnet));
        JSONObject json = new JSONObject(response);
        JSONObject transfer = json.optJSONObject("transfer");
        return transfer == null
                ? "Transfer added to put.io."
                : "Added: " + transfer.optString("name", "transfer");
    }

    public List<Transfer> transfers(String oauthToken) throws Exception {
        return parseTransfers(get(API_BASE + "/transfers/list", oauthToken, false));
    }

    public void cancelTransfer(String oauthToken, long transferId) throws Exception {
        requirePositiveId(transferId, "transfer");
        postForm("/transfers/cancel", oauthToken, "transfer_ids=" + transferId);
    }

    public FileListing files(String oauthToken, long parentId) throws Exception {
        String response = get(
                API_BASE + "/files/list?parent_id=" + parentId
                        + "&per_page=1000&mp4_status=true",
                oauthToken,
                false
        );
        return parseFiles(response, parentId);
    }

    public void deleteFile(String oauthToken, long fileId) throws Exception {
        requirePositiveId(fileId, "file");
        postForm("/files/delete", oauthToken, "file_ids=" + fileId);
    }

    public void renameFile(String oauthToken, long fileId, String name) throws Exception {
        requirePositiveId(fileId, "file");
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Enter a file name.");
        }
        postForm("/files/rename", oauthToken,
                "file_id=" + fileId + "&name=" + encode(name.trim()));
    }

    public String hlsStreamUrl(String oauthToken, long fileId) throws Exception {
        requireToken(oauthToken);
        requirePositiveId(fileId, "file");
        return API_BASE + "/files/" + fileId + "/hls/media.m3u8"
                + "?subtitle_key=all&oauth_token=" + encode(oauthToken.trim());
    }

    public String downloadUrl(String oauthToken, long fileId) throws Exception {
        requireToken(oauthToken);
        requirePositiveId(fileId, "file");
        return API_BASE + "/files/" + fileId + "/download"
                + "?oauth_token=" + encode(oauthToken.trim());
    }

    public void verifyToken(String oauthToken) throws Exception {
        get(API_BASE + "/account/info", oauthToken, false);
    }

    private static String get(String address, String token, boolean pendingAllowed) throws Exception {
        HttpURLConnection connection = (HttpURLConnection) URI.create(address).toURL().openConnection();
        connection.setConnectTimeout(15_000);
        connection.setReadTimeout(15_000);
        connection.setRequestProperty("Accept", "application/json");
        if (token != null) connection.setRequestProperty("Authorization", "Bearer " + token.trim());
        int status = connection.getResponseCode();
        String response = read(status >= 200 && status < 300
                ? connection.getInputStream()
                : connection.getErrorStream());
        connection.disconnect();
        if (pendingAllowed && (status == 400 || status == 404)) {
            throw new PendingAuthorization();
        }
        checkStatus(status, response);
        return response;
    }

    private static String postForm(String path, String token, String formBody) throws Exception {
        requireToken(token);
        HttpURLConnection connection = (HttpURLConnection) URI
                .create(API_BASE + path)
                .toURL()
                .openConnection();
        connection.setRequestMethod("POST");
        connection.setConnectTimeout(15_000);
        connection.setReadTimeout(30_000);
        connection.setDoOutput(true);
        connection.setRequestProperty("Authorization", "Bearer " + token.trim());
        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
        connection.setRequestProperty("Accept", "application/json");
        byte[] body = formBody.getBytes(StandardCharsets.UTF_8);
        try (OutputStream output = connection.getOutputStream()) {
            output.write(body);
        }
        int status = connection.getResponseCode();
        String response = read(status >= 200 && status < 300
                ? connection.getInputStream()
                : connection.getErrorStream());
        connection.disconnect();
        checkStatus(status, response);
        return response;
    }

    static List<Transfer> parseTransfers(String response) throws Exception {
        JSONArray array = new JSONObject(response).optJSONArray("transfers");
        if (array == null) {
            throw new IllegalStateException("put.io response was missing transfers.");
        }
        List<Transfer> transfers = new ArrayList<>();
        for (int index = 0; index < array.length(); index++) {
            JSONObject item = array.getJSONObject(index);
            transfers.add(new Transfer(
                    item.optLong("id"),
                    item.optLong("file_id"),
                    item.optString("name", "(unnamed transfer)"),
                    item.optString("status"),
                    item.optInt("percent_done"),
                    item.optString("error_message")
            ));
        }
        return Collections.unmodifiableList(transfers);
    }

    static FileListing parseFiles(String response, long requestedParentId) throws Exception {
        JSONObject root = new JSONObject(response);
        JSONArray array = root.optJSONArray("files");
        if (array == null) {
            throw new IllegalStateException("put.io response was missing files.");
        }
        List<FileItem> files = new ArrayList<>();
        for (int index = 0; index < array.length(); index++) {
            JSONObject item = array.getJSONObject(index);
            files.add(new FileItem(
                    item.optLong("id"),
                    item.optLong("parent_id"),
                    item.optString("name", "(unnamed file)"),
                    item.optString("content_type", "application/octet-stream"),
                    item.optString("file_type"),
                    item.optLong("size"),
                    item.optBoolean("is_mp4_available"),
                    item.optBoolean("need_convert")
            ));
        }
        JSONObject parent = root.optJSONObject("parent");
        long directoryId = parent == null ? requestedParentId
                : parent.optLong("id", requestedParentId);
        long parentDirectoryId = parent == null ? 0 : parent.optLong("parent_id");
        String name = parent == null
                ? (requestedParentId == 0 ? "put.io files" : "Folder")
                : parent.optString("name", "put.io files");
        return new FileListing(files, directoryId, parentDirectoryId, name);
    }

    private static void checkStatus(int status, String response) {
        if (status == 401 || status == 403) {
            throw new IllegalStateException("put.io rejected the token. Reconnect and try again.");
        }
        if (status < 200 || status >= 300) {
            String detail;
            try {
                detail = new JSONObject(response).optString("error_message", response);
            } catch (Exception ignored) {
                detail = response;
            }
            throw new IllegalStateException("put.io returned HTTP " + status + ": " + detail);
        }
    }

    private static void requireToken(String token) {
        if (token == null || token.trim().isEmpty()) {
            throw new IllegalStateException("Connect put.io before using account controls.");
        }
    }

    private static void requirePositiveId(long id, String kind) {
        if (id <= 0) {
            throw new IllegalArgumentException("A valid put.io " + kind + " is required.");
        }
    }

    private static String read(InputStream stream) throws Exception {
        if (stream == null) return "";
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            StringBuilder result = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) result.append(line);
            return result.toString();
        }
    }

    private static String encode(String value) throws Exception {
        return URLEncoder.encode(value, StandardCharsets.UTF_8.name());
    }

    public static final class DeviceCode {
        public final String code;

        public DeviceCode(String code) {
            this.code = code;
        }
    }

    public static final class Transfer {
        public final long id;
        public final long fileId;
        public final String name;
        public final String status;
        public final int percentDone;
        public final String errorMessage;

        Transfer(long id, long fileId, String name, String status,
                 int percentDone, String errorMessage) {
            this.id = id;
            this.fileId = fileId;
            this.name = name;
            this.status = status;
            this.percentDone = percentDone;
            this.errorMessage = errorMessage;
        }

        public boolean isDone() {
            return percentDone >= 100
                    || "DONE".equalsIgnoreCase(status)
                    || "COMPLETED".equalsIgnoreCase(status)
                    || "FINISHED".equalsIgnoreCase(status)
                    || "SEEDING".equalsIgnoreCase(status);
        }
    }

    public static final class FileItem {
        public final long id;
        public final long parentId;
        public final String name;
        public final String contentType;
        public final String fileType;
        public final long size;
        public final boolean mp4Available;
        public final boolean needsConversion;

        FileItem(long id, long parentId, String name, String contentType, String fileType,
                 long size, boolean mp4Available, boolean needsConversion) {
            this.id = id;
            this.parentId = parentId;
            this.name = name;
            this.contentType = contentType;
            this.fileType = fileType;
            this.size = size;
            this.mp4Available = mp4Available;
            this.needsConversion = needsConversion;
        }

        public boolean isDirectory() {
            return "application/x-directory".equalsIgnoreCase(contentType)
                    || "FOLDER".equalsIgnoreCase(fileType);
        }

        public boolean isVideo() {
            return !isDirectory() && (contentType.toLowerCase().startsWith("video/")
                    || "VIDEO".equalsIgnoreCase(fileType));
        }
    }

    public static final class FileListing {
        public final List<FileItem> files;
        public final long directoryId;
        public final long parentDirectoryId;
        public final String directoryName;

        FileListing(List<FileItem> files, long directoryId, long parentDirectoryId,
                    String directoryName) {
            this.files = Collections.unmodifiableList(new ArrayList<>(files));
            this.directoryId = directoryId;
            this.parentDirectoryId = parentDirectoryId;
            this.directoryName = directoryName;
        }
    }

    private static final class PendingAuthorization extends Exception {
    }
}

File: ios/PirateBrowser/PutIoService.swift
import Foundation

struct PutIoService {
    private let apiBase = URL(string: "https://api.put.io/v2")!

    func verifyToken(_ token: String) async throws {
        _ = try await HTTPClient.get(
            apiBase.appending(path: "account/info"),
            bearerToken: required(token)
        )
    }

    func addTransfer(token: String, magnet: String) async throws {
        _ = try await HTTPClient.postForm(
            apiBase.appending(path: "transfers/add"),
            token: required(token),
            values: ["url": magnet]
        )
    }

    func transfers(token: String) async throws -> [PutIoTransfer] {
        let data = try await HTTPClient.get(
            apiBase.appending(path: "transfers/list"),
            bearerToken: required(token)
        )
        let root = try HTTPClient.jsonObject(from: data)
        return (root["transfers"] as? [[String: Any]] ?? []).map { item in
            PutIoTransfer(
                id: number(item["id"]),
                name: text(item["name"], fallback: "Unnamed transfer"),
                status: text(item["status"], fallback: "UNKNOWN"),
                percentDone: decimal(item["percent_done"]),
                size: number(item["size"]),
                fileId: optionalNumber(item["file_id"])
            )
        }
    }

    func cancelTransfer(token: String, id: Int64) async throws {
        _ = try await HTTPClient.postForm(
            apiBase.appending(path: "transfers/cancel"),
            token: required(token),
            values: ["transfer_ids": String(id)]
        )
    }

    func files(token: String, parentId: Int64) async throws -> PutIoFileListing {
        var components = URLComponents(
            url: apiBase.appending(path: "files/list"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [URLQueryItem(name: "parent_id", value: String(parentId))]
        guard let url = components.url else { throw HTTPError.invalidURL }
        let data = try await HTTPClient.get(url, bearerToken: required(token))
        let root = try HTTPClient.jsonObject(from: data)
        let parent = (root["parent"] as? [String: Any]).map { number($0["id"]) } ?? parentId
        let items = (root["files"] as? [[String: Any]] ?? []).map { item in
            let contentType = text(item["content_type"])
            return PutIoFile(
                id: number(item["id"]),
                name: text(item["name"], fallback: "Unnamed file"),
                size: number(item["size"]),
                contentType: contentType,
                isDirectory: contentType == "application/x-directory"
                    || text(item["file_type"]).uppercased() == "FOLDER"
            )
        }
        return PutIoFileListing(
            parentId: parent,
            files: items.sorted {
                if $0.isDirectory != $1.isDirectory { return $0.isDirectory }
                return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
        )
    }

    func deleteFile(token: String, id: Int64) async throws {
        _ = try await HTTPClient.postForm(
            apiBase.appending(path: "files/delete"),
            token: required(token),
            values: ["file_ids": String(id)]
        )
    }

    func renameFile(token: String, id: Int64, name: String) async throws {
        _ = try await HTTPClient.postForm(
            apiBase.appending(path: "files/rename"),
            token: required(token),
            values: ["file_id": String(id), "name": name]
        )
    }

    func hlsURL(token: String, fileId: Int64) throws -> URL {
        var components = URLComponents(
            url: apiBase.appending(path: "files/\(fileId)/hls/media.m3u8"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "subtitle_key", value: "all"),
            URLQueryItem(name: "oauth_token", value: try required(token))
        ]
        guard let url = components.url else { throw HTTPError.invalidURL }
        return url
    }

    func downloadURL(token: String, fileId: Int64) throws -> URL {
        var components = URLComponents(
            url: apiBase.appending(path: "files/\(fileId)/download"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "oauth_token", value: try required(token))
        ]
        guard let url = components.url else { throw HTTPError.invalidURL }
        return url
    }

    func requestDeviceCode(clientId: String) async throws -> PutIoDeviceCode {
        let clientId = try required(clientId)
        var components = URLComponents(
            string: "https://api.put.io/v2/oauth2/oob/code"
        )!
        components.queryItems = [URLQueryItem(name: "app_id", value: clientId)]
        guard let url = components.url else { throw HTTPError.invalidURL }
        let root = try HTTPClient.jsonObject(from: try await HTTPClient.get(url))
        return PutIoDeviceCode(
            code: text(root["code"]),
            expiresIn: Int(number(root["expires_in"])),
            interval: max(1, Int(number(root["interval"])))
        )
    }

    func deviceToken(clientId: String, code: String) async throws -> String? {
        var components = URLComponents(
            string: "https://api.put.io/v2/oauth2/oob/code/\(code)"
        )!
        components.queryItems = [
            URLQueryItem(name: "app_id", value: try required(clientId))
        ]
        guard let url = components.url else { throw HTTPError.invalidURL }
        do {
            let root = try HTTPClient.jsonObject(from: try await HTTPClient.get(url))
            let token = text(root["oauth_token"])
            return token.isEmpty ? nil : token
        } catch HTTPError.rejected(let status, _) where status == 400 || status == 404 {
            return nil
        }
    }

    private func required(_ value: String) throws -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw HTTPError.rejected(400, "A put.io OAuth token is required.")
        }
        return trimmed
    }
}

private func text(_ value: Any?, fallback: String = "") -> String {
    if let value = value as? String, !value.isEmpty { return value }
    if let value = value as? NSNumber { return value.stringValue }
    return fallback
}

private func number(_ value: Any?) -> Int64 {
    if let value = value as? NSNumber { return value.int64Value }
    return Int64(text(value)) ?? 0
}

private func optionalNumber(_ value: Any?) -> Int64? {
    let result = number(value)
    return result > 0 ? result : nil
}

private func decimal(_ value: Any?) -> Double {
    if let value = value as? NSNumber { return value.doubleValue }
    return Double(text(value)) ?? 0
}
